### DISCUSSION TITLE
How to debug Symbol completion/creation?

### DISCUSSION THREAD

**Discussion by abgruszecki** [scala/scala3#11893] [Compiler internals Q&A] [answered]
I have a `Symbol` that seems like it shouldn't be created. How do I debug how it is created?

OR : I have a `Symbol` with a suspicious `.info` value. How do I debug how that value is calculated?

OR : I have a `Symbol` that asplodes during completion / isn't completed when it should be. How do I debug that?

**Comment by abgruszecki** [comment, marked as answer]
`Symbol`s have hash codes that are stable between compiler runs. To find out what creates a `Symbol`, first find out its hash code, then simply print a stack trace in `Symbol`'s constructor when the hash code matches. The stack trace will probably contain some [new*](https://github.com/lampepfl/dotty/blob/fix-11859-alt/compiler/src/dotty/tools/dotc/core/Symbols.scala#L527) methods in Symbols.scala.

When a symbol is created, its `.info` is set to a lazy type. This lazy type will later be used to actually calculate the `.info` for the symbol (see `complete` and `doComplete` in [SymbolLoader](https://github.com/lampepfl/dotty/blob/fix-11859-alt/compiler/src/dotty/tools/dotc/core/SymbolLoaders.scala#L313)). One way to debug information about the completion of the symbol is to print information when the symbol with the correct hash code is being completed.

**Reply by smarter** [reply]
> Symbols have hash codes 

Right, `hashCode` for a symbol delegates to `id`, which is shorter to type :). And you can get the compiler to append symbol ids in its pretty-printed output using `-uniqid`

> then simply print a stack trace

Which can be done using `Thread.dumpStack`

> debug information about the completion of the symbol

You can set `showCompletions` to true in https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/config/Config.scala to get a log of every completion, where nested completions will appear nested in the output (pretty useful for debugging cycles).
